contracts/secret-storage/SecretStorage.sol

```diff
@@ -0,0 +1,914 @@
+        uint256 _secretId,
+        uint256 _signTimestamp,
+        bytes memory _signature,
+        address _owner
+    ) internal {
+        address enclaveAddress = _verifyAcknowledgementSign(_secretId, _owner, _signTimestamp, _signature);
+
+        uint256 enclaveIndex = _getSelectedEnclaveIndex(_secretId, enclaveAddress);
+        userStorage[_secretId].selectedEnclaves[enclaveIndex].hasAcknowledgedStore = true;
+
+        emit EnclaveAcknowledgedStore(_secretId, enclaveAddress);
+    }
+
+    function _verifyAcknowledgementSign(
+        uint256 _secretId,
+        address _owner,
+        uint256 _signTimestamp,
+        bytes memory _signature
+    ) internal view returns(address signer) {
+        if (block.timestamp > _signTimestamp + ATTESTATION_MAX_AGE)
+            revert SecretStorageSignatureTooOld();
+
+        bytes32 hashStruct = keccak256(abi.encode(ACKNOWLEDGE_TYPEHASH, _secretId, _owner, _signTimestamp));
+        bytes32 digest = keccak256(abi.encodePacked("\x19\x01", DOMAIN_SEPARATOR, hashStruct));
+        signer = digest.recover(_signature);
+
+        _allowOnlyVerified(signer);
+    }
+
+    function _getSelectedEnclaveIndex(
+        uint256 _secretId,
+        address _enclaveAddress
+    ) internal view returns (uint256) {
+        uint256 len = userStorage[_secretId].selectedEnclaves.length;
+        for (uint256 index = 0; index < len; index++) {
+            if(userStorage[_secretId].selectedEnclaves[index].enclaveAddress == _enclaveAddress)
+                return index;
+        }
+        revert SecretStorageEnclaveNotFound();
+    }
+
+    function _acknowledgeStoreFailed(
+        uint256 _secretId
+    ) internal {
+        if(block.timestamp <= userStorage[_secretId].startTimestamp + ACKNOWLEDGEMENT_TIMEOUT)
+            revert SecretStoreAcknowledgementTimeoutPending();
+
+        bool ackFailed;
+        uint256 len = userStorage[_secretId].selectedEnclaves.length;
+        for (uint256 index = 0; index < len; index++) {
+            if(!userStorage[_secretId].selectedEnclaves[index].hasAcknowledgedStore) {
+                ackFailed = true;
+                break;
+            }
+        }
+
+        if(!ackFailed)
+            revert SecretStoreAcknowledgedAlready();
+
+        address owner = userStorage[_secretId].owner;
+        uint256 usdcDeposit = userStorage[_secretId].usdcDeposit;
+        delete userStorage[_secretId];
+
+        USDC_TOKEN.safeTransfer(owner, usdcDeposit);
+
+        emit EnclaveAcknowledgementFailed(_secretId);
+    }
+
+    function _markStoreAlive(
+        uint256 _secretId,
+        uint256 _signTimestamp,
+        bytes memory _signature,
+        address _owner
+    ) internal {
+        address enclaveAddress = _verifyStoreAliveSign(_secretId, _owner, _signTimestamp, _signature);
+
+        uint256 enclaveIndex = _getSelectedEnclaveIndex(_secretId, enclaveAddress);
+        userStorage[_secretId].selectedEnclaves[enclaveIndex].lastAliveTimestamp = block.timestamp;
```

Comment:
we should use enclave timestamp here, otherwise alive time duration will depend on when the transaction is getting include. Although, clocks can be out-of-sync for blockchain and enclave. Will address that later and how to tackle it.